torchgeo/datasets/mmflood.py

```diff
@@ -0,0 +1,422 @@
+
+        Raises:
+            IndexError: if query is not found in the index
+        """
+        hits = self.index.intersection(tuple(query), objects=True)
+        indexes = cast(list[int], [hit.id for hit in hits])
+
+        if not indexes:
+            raise IndexError(
+                f'query: {query} not found in index with bounds: {self.bounds}'
+            )
+
+        image = self._load_image(indexes, query)
+        mask = self._load_target(indexes, query)
+
+        sample = {'image': image, 'mask': mask, 'crs': self.crs, 'bounds': query}
+
+        if self.transforms is not None:
+            sample = self.transforms(sample)
+
+        return sample
+
+    def __len__(self) -> int:
+        """Return the number of data points in the dataset.
+
+        Returns:
+            length of the dataset
+        """
+        return len(self.folders)
+
+    def _load_metadata(self) -> pd.DataFrame:
+        """Load metadata.
+
+        Returns:
+            dataframe containing metadata
+        """
+        df = pd.read_json(
+            os.path.join(self.root, self.metadata['metadata_file'])
+        ).transpose()
+        return df
+
+    def _load_folders(self, check_folders: bool = False) -> list[dict[str, str]]:
+        """Load folder paths.
+
+        Args:
+            check_folders: if True, verify pairings of all s1, dem and mask data across all the folders
+
+        Returns:
+            list of dicts of s1, dem and masks folder paths
+        """
+        # initialize tif file lists containing masks, DEM and S1_raw data
+        folders = self.metadata_df[
+            self.metadata_df['subset'] == self.split
+        ].index.tolist()
+
+        image_files = []
+        mask_files = []
+        dem_files = []
+        for f in folders:
+            path = os.path.join(self.root, self.metadata['directory'], f'{f}-*')
+            image_files += glob(os.path.join(path, 's1_raw', '*.tif'))
+            mask_files += glob(os.path.join(path, 'mask', '*.tif'))
+            dem_files += glob(os.path.join(path, 'DEM', '*.tif'))
+
+        image_files = sorted(image_files)
+        mask_files = sorted(mask_files)
+        dem_files = sorted(dem_files)
+
+        # Verify image, dem and mask lengths
+        assert (
+            len(image_files) > 0
+        ), f'No images found, is the given path correct? ({self.root!s})'
+        assert (
+            len(image_files) == len(mask_files)
+        ), f'Length mismatch between tiles and masks: {len(image_files)} != {len(mask_files)}'
+        assert len(image_files) == len(
+            dem_files
+        ), 'Length mismatch between tiles and DEMs'
+
+        res_folders = [
+            {'s1_raw': img_path, 'mask': mask_path, 'dem': dem_path}
+            for img_path, mask_path, dem_path in zip(image_files, mask_files, dem_files)
+        ]
+
+        if not check_folders:
+            return res_folders
+
+        # Verify image, dem and mask pairings
```

Comment:
Could this verification instead be part of  `self._verify()` ? So first check the existence of the necessary files and then also compute this check?